lib/elixir/lib/code.ex

```diff
@@ -206,6 +206,237 @@ defmodule Code do
+      # It is empty
+      {[], _} ->
+        :expr
+
+      # Start of a dot or alias
+      {[?. | rest], _} ->
+        case identifier_to_autocomplete(rest) do
+          {:alias, prev} -> {:alias_or_dot, prev}
+          {:local_or_var, prev} -> {:dot, {:var, prev}, []}
+          {:unquoted_atom, _} = prev -> {:dot, prev, []}
+          {:dot, _, _} = prev -> {:dot, prev, []}
+          _ -> :none
+        end
+
+      # It is a local or remote call with parens
+      {[?( | rest], _} ->
+        call_to_autocomplete(rest)
+
+      # A local arity definition
+      {[?/ | rest], _} ->
+        case identifier_to_autocomplete(rest) do
+          {:local_or_var, acc} -> {:local_arity, acc}
+          {:dot, base, acc} -> {:dot_arity, base, acc}
+          _ -> :none
+        end
+
+      # Starting a new expression
+      {[h | _], _} when h in @operators_and_non_closing_puctuation ->
+        :expr
+
+      # It is a local or remote call without parens
+      {rest, spaces} when spaces > 0 ->
+        call_to_autocomplete(rest)
+
+      # It is an identifier
+      _ ->
+        identifier_to_autocomplete(reverse)
+    end
+  end
+
+  defp strip_spaces([h | rest], count) when h in @space, do: strip_spaces(rest, count + 1)
+  defp strip_spaces(rest, count), do: {rest, count}
+
+  defp call_to_autocomplete(reverse) do
+    case identifier_to_autocomplete(reverse) do
+      {:local_or_var, acc} -> {:local_call, acc}
+      {:dot, base, acc} -> {:dot_call, base, acc}
+      _ -> :none
+    end
+  end
+
+  defp identifier_to_autocomplete(reverse) do
+    case identifier(reverse) do
+      # Parse :: first to avoid ambiguity atoms
+      {:alias, false, '::' ++ _, _} -> :none
+      {kind, _, '::' ++ _, acc} -> alias_or_local_or_var(kind, acc)
+      # Now handle atoms, any other atom is unexpected
+      {_kind, _, ':' ++ _, acc} -> unquoted_atom(acc)
+      {:atom, _, _, _} -> :none
+      # Parse .. first to avoid ambiguity with dots
+      {:alias, false, _, _} -> :none
+      {kind, _, '..' ++ _, acc} -> alias_or_local_or_var(kind, acc)
+      # Everything else
+      {kind, _, '.' ++ rest, acc} -> alias_or_dot(kind, rest, acc)
+      {kind, _, _, acc} -> alias_or_local_or_var(kind, acc)
+      :none -> :none
+    end
+  end
+
+  defp alias_or_dot(kind, rest, acc) do
+    case {kind, identifier_to_autocomplete(rest)} do
+      {:alias, {:alias, prev}} -> {:alias, prev ++ '.' ++ acc}
+      {:identifier, {:local_or_var, prev}} -> {:dot, {:var, prev}, acc}
+      {:identifier, {:unquoted_atom, _} = prev} -> {:dot, prev, acc}
+      {:identifier, {:alias, _} = prev} -> {:dot, prev, acc}
+      {:identifier, {:dot, _, _} = prev} -> {:dot, prev, acc}
+      _ -> :none
+    end
+  end
+
+  defp alias_or_local_or_var(:alias, acc), do: {:alias, acc}
+  defp alias_or_local_or_var(:identifier, acc), do: {:local_or_var, acc}
+  defp alias_or_local_or_var(_, _), do: :none
+
+  defp unquoted_atom(acc), do: {:unquoted_atom, acc}
```

Comment:
Not necessary?

Reply:
Most of the parser code works with charlists because it is simpler to work with codepoints. Given the result of this will likely be piped into the parser, such as Code.string_to_quoted, I chose to return a charlist to avoid double or triple conversion.